Preliminary definitions for n5 metadata

virginiascarlett opened the pull request:
Hey @cmhulbert and @bogovicj, 

I've written up some definitions for the metadata for the n5-viewer format and I'd love if you could take a look. I made some opinionated choices, so don't just trust me!

BTW, let me know if there's an ideal way to describe the relationship between these formats. E.g. is bigcat a "flavor" of n5? Can I just use "n5 format" interchangeably with "n5-viewer" format?

Looking ahead, once the metadata are largely done, I'll work on schematizing hierarchy configurations next.

Thanks in advance!

bogovicj commented:
I'd prefer "n5-viewer" format to "n5" format, but let me look through this and sleep on it.
Thanks @virginiascarlett !

bogovicj commented:
This was written very clearly and thoroughly, thanks again @virginiascarlett 

I'm tempted to split this page into two or three related pages, but whether that makes sense depends on how we choose to organize things. For example

* `n5-array` (the metadata required of an array stored in n5)
   * `n5-viewer`  (the additional requirements of the n5-viewer metadata standard)
   * `bigcat` (the additional requirements of the bigcat metadata standard)
  
where `n5-array` would hold `dimensions`, `blockSize`, `compression` and is the "base" required metadata for the two others.  i.e. to conform to the `n5-viewer` standard a dataset must conform to `n5-array` standard (perhaps with some more restrictions - like on the allowable length of `dimensions`, for example). It may make sense to  merge `n5-viewer` and `bigcat` into one document, but I'm not sure yet.

cmhulbert commented on n5-viewer/README.md:
```diff
@@ -1,2 +1,78 @@
+
+A couple of requirements are not currently enforced by the schema:
+1. The order of dimensions must be xyzt (or xyz, xyt, or xy). Dimension order is assumed by the software and is not expressed anywhere in the metadata, so a metadata document with incorrect dimension order will pass schema validation. 
+2. Channel is expressed in the hierarchy structure, not the metadata. Folders for each channel must be named e.g. c0/, c1/. 
```
@bogovicj is it true that channel names are required to be named `c0`...`cN`? I know this is convention, at least

cmhulbert commented on n5-viewer/README.md:
```diff
@@ -1,2 +1,78 @@
+2. Average downsampling, in which subsets of samples are averaged to produce a new, smaller array. For example, downsamplingFactors of 4 means every four samples were averaged to produce a new sample. If dimensions are downsampled evenly, the total number of samples is reduced by a factor of N^M, where N is the number of dimensions and M is the downsampling factor.   
+
+A few requirements are not currently (or not yet) enforced by the schema:
+1. s0/ should have `downsamplingFactor`s of 1 (e.g. [1,1] for a 2-dimensional image), since s0 should be the original image. 
```
I agree it would be nice for consistency for `s0` to have `downsamplingFactor` of `[1,...,1]`. You mention it's not enforced now. I think in practice it is usually not present at all for `s0`. Not sure if we want to claim that it should be, or not.

bogovicj replied:
I'm also not sure, I don't really care.  

The small part of me that does care likes that it's not present for `s0`.  Having the parameter there implies that it could have been something other than `[1...,1]`, and I don't like that.

cmhulbert commented:
Thanks @virginiascarlett, this is something that I think has been missing for a while!
I agree with @bogovicj about splitting some of this out. My understanding is that there really is no standard "n5 format" except for the array metadata (and a version tag!). The rest are conventions that are built around that, that are used with less consistency than would be ideal. In particular, the resolution is ambiguous, and I think you correctly define what `resolution`, `pixelResolution` and `downsamplingFactors` represent, but they are I think not all always necessary, since you can infer some from the others.

Something else that I think would be helpful is to separate out the metadata requirements for a single scale dataset (either standalone, or as a child of a multiscale dataset, say `s0`) from the multiscale metadata requirements. For example, `resolution` and `pixelResolution` are helpful for single scale, while `downsamplingFactors` is helpful to know for multiscale (in conjunction with the base `resolution` of `s0`)

virginiascarlett commented:
Apologies for taking forever with this, and thanks for the feedback. I think I have resolved all of your concerns, except Caleb's comment about channel names. I split up the metadata fields into "required", "optional", and "conditionally required", and updated the descriptions of resolution, pixelResolution, and downsamplingFactors to reflect some situations where they may/may not be required.

I will tidy up the schemas accordingly, so that it's a little clearer that everything builds off of the "base" n5 array, but I will do that outside of this PR, once this PR is resolved.
